File: merge-intervals/merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (i1, i2) -> Integer.compare(i1[0], i2[0]));
        
        ArrayList<int[]> res = new ArrayList<>();
        int[] prevInterval = intervals[0];
        for (int i = 1; i < intervals.length; i++) {
            int[] interval = intervals[i];
            if (interval[0] <= prevInterval[1]) {
                prevInterval[1] = Math.max(prevInterval[1], interval[1]);
            } else {
                res.add(prevInterval);
                prevInterval = interval;
            }
        }
        res.add(prevInterval);
        
        int[][] ans = new int[res.size()][2];
        for (int i = 0; i < res.size(); i++) {
            ans[i] = res.get(i);
        }
        
        return ans;
    }
}

File: merge-intervals/merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key=lambda i:i[0])
        
        ans = [intervals[0]]
        
        for i in range(1, len(intervals)):
            prev = ans[-1]
            curr = intervals[i]
            
            if prev[1] >= curr[0]:
                prev[1] = max(prev[1], curr[1])
            else:
                ans.append(curr)
            
        return ans